### ISSUE TITLE
Which tool for front-end javascript?

### ISSUE THREAD

**Issue by SimonLab** [charterindex/application#63]
For "enhanced" user experience on the website I was thinking of implementing some features (ex: autocomplete) with [Elm](http://elm-lang.org/). Elm is a functional programming language which make it easier to track the states of the application and it has a good compiler which give indication on errors and how to fix them which makes debugging easier too. I think that are the two main points why we should try Elm but there are a lot more useful aspects (ref: https://github.com/dwyl/learn-elm) compare to just use vanilla javascript.

@nelsonic @iteles  do you have thoughts or reservations on why we shouldn't use Elm?

**Comment by nelsonic** [comment]
@SimonLab `elm-lang` was my _suggestion_ (_too_) when speaking to Chris & Peter _originally_.

(_We are using elm for a _few_ of our clients and it's still my **preference** for obvious reasons: **reliability, simplicity & speed**. The learning curve for a **complete beginner** is only a few hours and the benefits are immediate in terms of lower "cost" in bug-fixing and extending apps..._)

There was a _mention_ of Angular.js however I feel that the _complexity_ of Angular 
(_Typescript, RxJS & "Bring-your-own" architecture_) is more than we _need_ for this project.

I think the best way to _demonstrate_ this is for a small _feature_ demo.
Do you have a _user-story_ in mind for what you would implement client-side?

**Comment by SimonLab** [comment]
My first idea is to try to use Elm for the autocomplete search on the home page. However as we need some more data from the API this might not be the smallest feature to start to implement
see #59 but we can still mock some yacht names.

**Comment by nelsonic** [comment]
@SimonLab the place I would _suggest_ using `Elm` would be in the charter request form.
_However_ to achieve AMP-compliance #25 we will need to be "smart" about this 
by having a "Plain HTML" form which, when _submitted_ is "upgraded" to an `Elm` form.

@Cleop could you direct us to the User Story for the "Enquiry Form"? (_I can't see it in the backlog ..._)

**Comment by nelsonic** [comment]
@SimonLab the place I would _suggest_ using `Elm` would be in the charter request form.
_However_ to achieve AMP-compliance #25 we will need to be "smart" about this 
by having a "Plain HTML" form which, when _submitted_ is "upgraded" to an `Elm` form.

In light of the fact that the Search Bar (_with "autocompletion"_) will be on the "Landing Page"
We will need to build the "search" functionality as a "Stand Alone" page which we will then include in an `<amp-iframe>` see: https://stackoverflow.com/a/36269075/1148249 (_in order to be AMP-compliant_)

This is actually a _good_ thing because it means we _develop_ the "interactive" functionality 
in isolation from any "content" and then include it dynamically wherever it's needed.

**Comment by SimonLab** [comment]
We are working on the "where" dropdown: #76 and specifically on the pillbox when an area is selected:

![image](https://user-images.githubusercontent.com/6057298/38424145-ffeac46c-39a7-11e8-9566-06f4aeb7b882.png)

One way to do this is by creating some vanilla functions which will listen to any "change" events on the dropdown and then display or hide the pillbox and the "where" dropdown, something like:
```
function generatePill() {
  var where = document.getElementById("where");

  where.addEventListener("change", function(e) {
    var selectedOption = e.target.options[e.target.selectedIndex].text;
    var selectedValue = e.target.options[e.target.selectedIndex].value;
    var pill = document.getElementById("where-pill");
    if (selectedValue == "") {
      pill.classList.remove("dib");
      pill.classList.add("dn");
    } else {
      pill.classList.remove("dn");
      pill.classList.add("dib");
      pill.innerHTML = selectedOption;
    }
  });
}

generatePill();
```

From past experience I think this kind of functions are a bit difficult to maintain and keep them without bugs. We will have more feature to implement with javascript and the code base will get bigger with more place for side effects. 

I think Elm might be a good solution to try on the "where" dropdown and pillbox and will help us to manage all the states of the application and avoid side effects. @nelsonic @PeterTrotter do you have any thoughts on this and would it be a good case to try Elm on the application?

**Comment by PeterTrotter** [comment]
Elm looks pretty cool but I think it's more important for CI in general to keep required maintenance knowledge to a minimum. Each new language makes it just a little harder to get new developers in.

I think I'm gonna go have a little tinker with Elm when I get a chance.

**Comment by nelsonic** [comment]
@PeterTrotter, we think you will be _pleasantly_ surprised by Elm's simplicity/elegance.

A few useful links:

+ After React, Elm?: https://blog.wraptime.io/after-react-elm-152bd6559cb1
+ Why Elm is going to change the world: https://hackernoon.com/why-elm-is-going-to-change-the-world-f5a6c693b2ca
+ Elm Architecture tutorial in JavaScript: https://github.com/dwyl/learn-elm-architecture-in-javascript
+ **Official Guide**: https://guide.elm-lang.org (_with all examples editable online_)
+ eBook: https://dwyl.s3.amazonaws.com/Elm_in_Action_v5_MEAP-2.pdf (_"in-progress" book written by one of the most respected Elm experts, Richard Feldman_)
  + Sample code (_for the eBook_): https://github.com/nelsonic/photo-groove
+ London Meetup: https://www.meetup.com/Elm-London-Meetup (_700+ members, great for learning from people with lots of experience._)

## tl;dr

Someone `new` to the team will _always_ have to learn (_some part_) of any "framework" selected
in order to maintain/update/extend the CI App so it will take _some_ time to get up-to-speed.  

At _first_ I was "reluctant" to use a "new" language because I thought  that anyone _joining_ one of our teams would take a "long time" to _learn_ `before` they could be "productive" ...
But once I saw how _tiny_ Elm is, (_much less to learn than `ES6` + `immutable.js`, `Typescript` etc._)
and how it's "_impossible_" for a new team member to "break" an existing app (_because the compiler won't let the App be deployed if anything is "broken"_) it was _easy_ to use it for Apps.
e.g: https://apps.qa.tcmoney.io (_ThomasCook Money's new Mobile-first Web Apps - built with Elm - which go "live" at the end of this month_)

The key distinction is that Elm is "self-contained"; the language includes the "architecture", (_super beginner-friendly_) compiler and run-time with debugger. 

Imagine that you (_Charter Index_) need to ***embed*** a "**Yacht Finder Widget**" in any website 
and need it work _flawlessly_ with _minimal_ effort, Elm will make this possible in a **bulletproof** way with _much_ less effort (_fewer lines of code that load faster_) than _any_ other "framework".

**Comment by PeterTrotter** [comment]
@nelsonic sent you an email requesting a quick call regarding this issuae and #79

**Comment by nelsonic** [comment]
@PeterTrotter thank you for the Hangout and confirming the "green light" to use Elm for front-end. 🎉 
@SimonLab / @Cleop if you have time for a Hangout at some point (_e.g: Lunch_) please LMK. 🕧

**Comment by iteles** [comment]
Closing this issue. As per the above, we went ahead with Elm.
